File: tests/assets/algorithms-demo/pointer-example.cpp


class Number
{
public:
    Number(int value) : value(value) {}

    int get() const
    {
        return value;
    }

private:
    int value;
};

class One : public Number
{
public:
    One() : Number(1) {}
};

class Two : public Number
{
public:
    Two() : Number(2) {}
};

class C
{
public:
    static void main()
    {
        C *c = new C(); // o_l_34 (o_3)
        (*c).m();
    }

    void m()
    {
        One *n1 = new One(); // o_12
        Two *n2 = new Two(); // o_13
        Number *x = (*this).id(n1);
        Number *y = (*this).id(n2);
        (*x).get();
    }

    Number *id(Number *n)
    {
        return n;
    }
};

int main()
{
    C::main();
}

File: tests/assets/algorithms-demo/pointer-example.py
class C:
    @staticmethod
    def main():
        c = C()  # o_3
        c.m()

    def m(self):
        n1 = One()  # o_12
        n2 = Two()  # o_13
        x = self.id(n1)
        y = self.id(n2)
        x.get()

    def id(self, n: "Number") -> "Number":
        return n


class Number:
    def __init__(self, value):
        self.value = value

    def get(self):
        return self.value


class One(Number):
    def __init__(self):
        super().__init__(1)


class Two(Number):
    def __init__(self):
        super().__init__(2)
